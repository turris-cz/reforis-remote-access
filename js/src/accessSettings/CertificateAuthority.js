/*
 * Copyright (C) 2019-2021 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button,
    useAlert,
    useAPIDelete,
    API_STATE,
    formFieldsSize,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({
    authority,
    accessEnabled,
    onSuccess,
}) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority."));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <div className={formFieldsSize}>
            <h2>{_("Certificate Authority")}</h2>
            {accessEnabled ? (
                <p>
                    {_(
                        "You can't delete certificate authority while remote access is enabled. To delete it, you need to disable the access first."
                    )}
                </p>
            ) : (
                <>
                    <p>
                        {_(
                            "Your certificate authority is set up properly. Please note that if you delete it, all clients will have their access revoked."
                        )}
                    </p>
                    <div className="text-right">
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete Certificate Authority")}
                        </Button>
                    </div>
                </>
            )}
        </div>
    );
}
