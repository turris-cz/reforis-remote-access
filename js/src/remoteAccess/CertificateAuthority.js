/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    Button, useAlert, useAPIDelete, API_STATE,
} from "foris";

import API_URLs from "API";

export const CA_STATUS = {
    MISSING: "missing",
    READY: "ready",
    GENERATING: "generating",
};

CertificateAuthority.propTypes = {
    authority: PropTypes.object.isRequired,
    accessEnabled: PropTypes.bool.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

export default function CertificateAuthority({ authority, accessEnabled, onSuccess }) {
    const [setAlert] = useAlert();

    const [deleteResponse, deleteCA] = useAPIDelete(API_URLs.authority);
    useEffect(() => {
        if (deleteResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        } else if (deleteResponse.state === API_STATE.ERROR) {
            setAlert(_("Cannot delete certificate authority"));
        }
    }, [deleteResponse, onSuccess, setAlert]);

    if (authority.status !== CA_STATUS.READY) {
        return null;
    }

    return (
        <>
            <h3>{_("Certificate Authority")}</h3>
            {accessEnabled
                ? <p>{_("You can't delete certificate authority while remote access is enabled. In order to delete it you need to disable the access first.")}</p>
                : (
                    <>
                        <p>{_("Your certificate authority is set up properly. Please note that if you delete it all clients will have their access revoked.")}</p>
                        <Button
                            onClick={() => deleteCA()}
                            loading={deleteResponse.state === API_STATE.SENDING}
                            forisFormSize
                        >
                            {_("Delete certificate authority")}
                        </Button>
                    </>
                )}
        </>
    );
}
