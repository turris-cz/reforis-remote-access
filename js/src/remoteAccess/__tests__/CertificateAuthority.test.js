/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import mockAxios from 'jest-mock-axios';
import { WebSockets, API_STATE } from "foris";
import { render, getByText, wait, fireEvent } from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";

import CertificateAuthority, { CA_STATUS } from "../CertificateAuthority";

describe("<CertificateAuthority />", () => {
    const webSockets = new WebSockets();
    const onSucces = jest.fn();

    function renderComponent(apiState, authority = {}) {
        return render(
            <CertificateAuthority
                apiState={apiState}
                ws={webSockets}
                authority={authority}
                onSuccess={onSucces}
            />
        );
    }

    describe("authority is ready", () => {
        let container;

        beforeEach(() => {
            ({ container } = renderComponent(
                API_STATE.SUCCESS,
                { status: CA_STATUS.READY },
            ));
        });

        it("should render component", () => {
            expect(container).toMatchSnapshot();
        });

        it("should send request when button is clicked", () => {
            fireEvent.click(getByText(container, "Delete CA"));
            expect(mockAxios.delete).toBeCalledWith("/reforis/subordinates/api/authority", expect.anything());
        });

        it("should handle error", async () => {
            fireEvent.click(getByText(container, "Delete CA"));
            mockJSONError();
            await wait(() => {
                expect(mockSetAlert).toHaveBeenCalledWith("Cannot delete certificate authority");
            });
        });

        it("should handle success", async () => {
            fireEvent.click(getByText(container, "Delete CA"));
            mockAxios.mockResponse({});
            await wait(() => {
                expect(onSucces).toHaveBeenCalledTimes(1);
            });
        });
    });

    it("should render missing component", () => {
        const { container } = renderComponent(
            API_STATE.SUCCESS,
            { status: CA_STATUS.MISSING },
        );
        expect(container).toMatchSnapshot();
    });

    it("should render spinner", () => {
        const { container } = renderComponent(API_STATE.SENDING);
        expect(container).toMatchSnapshot();
    });

    it("should render error", () => {
        const { container } = renderComponent(API_STATE.ERROR);
        expect(container).toMatchSnapshot();
    });
});
