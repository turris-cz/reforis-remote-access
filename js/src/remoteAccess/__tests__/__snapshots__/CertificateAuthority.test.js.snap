// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CertificateAuthority /> authority is ready should render component 1`] = `
<div>
  <h3>
    Certificate Authority
  </h3>
  <p>
    Your certificate authority (CA) is set up properly. Please note that if you delete it all clients will have their access revoked.
  </p>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    type="button"
  >
     
     
    Delete CA
  </button>
</div>
`;

exports[`<CertificateAuthority /> should render error 1`] = `
<div>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`<CertificateAuthority /> should render missing component 1`] = `
<div>
  <h3>
    No certification authority
  </h3>
  <p>
    Currently there is no certificate authority (CA) dedicated to remote access. A CA is required to generate access tokens to authenticate. To proceed you need to generate it first.
  </p>
  <button
    class="btn btn-primary  offset-lg-8 col-lg-3 col-sm-12"
    type="button"
  >
     
     
    Generate CA
  </button>
</div>
`;

exports[`<CertificateAuthority /> should render spinner 1`] = `
<div>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border "
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
    <div
      class="spinner-text"
    />
  </div>
</div>
`;
